### PR TITLE
Updates and elaboration on packaging and distribution.

### PR DESCRIPTION
**Pull request by eirrgang**
Updates for PEP 517 in accordance with https://github.com/MolSSI/cookiecutter-cms/pull/150 and related updates to the cookiecutter.

Cite resources on packaging and distributing for PyPI, with some additional notes.

### CODE REVIEW

**Comment by eirrgang** [comment]
Note: @Lnaden and @janash think that we will probably go all the way to PEP 621 setuptools configuration, so some of the references to setup.cfg should be updated wrt pyproject.toml

File: _episodes/09-documentation.md
Comment (eirrgang) [inline _episodes/09-documentation.md:617]:
```suggestion
Update your `README.md` and `setup.cfg` files with your new ReadTheDocs URL.
```

File: _episodes/03-github.md
Comment (eirrgang) [inline _episodes/03-github.md:82]:
```suggestion
For now, at least, update your `pyproject.toml`, and put this GitHub repository URL
```

File: _episodes/01-package-setup.md
Comment (eirrgang) [inline _episodes/01-package-setup.md:361]: This is irrelevant if we are moving setuptools config to pyproject.toml
```suggestion
```

File: _episodes/01-package-setup.md
Comment (eirrgang) [inline _episodes/01-package-setup.md:351]:
```suggestion
Two of the files CookieCutter generated are `pyproject.toml` and `setup.cfg`.
These are the configuration files for our packaging and testing tools.
```

File: _episodes/01-package-setup.md
Comment (eirrgang) [inline _episodes/01-package-setup.md:352]:
```suggestion
`pyproject.toml` tells [setuptools] about your package (such as the name and version) as well as which code files to include.
```

File: _episodes/01-package-setup.md
Comment (eirrgang) [inline _episodes/01-package-setup.md:348]:
```suggestion
#### Reviewing the generated config files
```

**Review by eirrgang — COMMENTED** [review]
Updates that assume we are embracing PEP 621.

Corresponding updates to lesson 10 are being made first in the hackmd doc.

File: _episodes/10-distribution.md
Comment (janash) [inline _episodes/10-distribution.md:112]: I want to have a discussion on this, so I don't think this is beyond the scope.

**Review by janash — CHANGES_REQUESTED** [review]
I'd like to go through the distribution lesson a bit more. It's not clear to me right now which points are informational and which are tutorial (what steps should be executed when). Would like to keep hands on (test-pypi) upfront, then we can put information about conda at the end.

**Comment by eirrgang** [comment]
> I'd like to go through the distribution lesson a bit more. It's not clear to me right now which points are informational and which are tutorial (what steps should be executed when). Would like to keep hands on (test-pypi) upfront, then we can put information about conda at the end.

Just a reminder: I have mostly refrained from updating this branch since beginning the working draft on hackmd.

I'm still waiting for updates/feedback there, but I will import a snapshot to this PR for clarity and convenience.

File: _episodes/10-distribution.md
Comment (eirrgang) [inline _episodes/10-distribution.md:25]:
```suggestion
Create an access token, and *don't lose it!*
(I keep a copy of the text for my token in my password manager.)

The access token will be a string that begins with `pypi-` followed by a sequence of base64 encoded bytes
that we will refer to as "YOURTOKEN" below.

Create `$HOME/.pypirc`. It should look something like the following, substituting your access token in the password field.

~~~
[testpypi]
  username = __token__
  password = YOURTOKEN
```

File: _episodes/10-distribution.md
Comment (eirrgang) [inline _episodes/10-distribution.md:162]:
```suggestion
Read through https://packaging.python.org/en/latest/tutorials/packaging-projects/
for a build-system-agnostic tutorial on
* packaging Python projects for distribution
* publishing packages to pypi.org for automatic download with `pip install`

By the end of the [tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/),
you will understand how to get an account on `pypi.org`,
build a distribution archive with `python -m build`,
and use twine to upload your package.

The [tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/) should go quickly because,
as you will note, most of the configuration it describes
has already been pre-configured by the cookiecutter.
```

**Comment by janash** [comment]
@eirrgang - can you share the hackmd for this with me?

**Comment by eirrgang** [comment]
It's gone. I consolidated to this PR and removed the hackmd page to avoid
crossed wires. :-)

File: _episodes/10-distribution.md
Comment (eirrgang) [reply _episodes/10-distribution.md:25]: See https://github.com/MolSSI-Education/python-package-best-practices/pull/80#discussion_r959411486

File: _episodes/10-distribution.md
Comment (eirrgang) [reply _episodes/10-distribution.md:162]: See https://github.com/MolSSI-Education/python-package-best-practices/pull/80#discussion_r959417100

File: _episodes/10-distribution.md
Comment (eirrgang) [inline _episodes/10-distribution.md:48]: The conclusion at the last meeting was that we should briefly address the existence of these features. I will add a paragraph or two. Do we think an informational call-out would be sufficient? What does that look like? `{: .callout}`?

Comment (janash) [reply _episodes/10-distribution.md:48]: Yes, I think `{: .callout}` would be appropriate.

**Review by janash — COMMENTED** [review]
I think this is a good first pass. Let's add a callout at the top that says this lesson is preliminary and under development. We can build it out more in the review next year. Thanks Eric!

File: _episodes/10-distribution.md
Comment (eirrgang) [inline _episodes/10-distribution.md:48]:
```suggestion
want to include in a text file or in a blurb in your GitHub
[Releases tab](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases).
This can be a good place to publish [release notes](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes).
```

File: _episodes/10-distribution.md
Comment (eirrgang) [inline _episodes/10-distribution.md:69]:
```suggestion

~~~
Tags created in your local repository are not automatically shared when you `git push` a branch.
You can push tags the same way you push branches, or you can create the tag directly on GitHub.
With the GitHub interface, you can associate tags with
[Releases](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases)
to provide a place to publish release notes and automatically generate links to archive files of
your repository at the tagged commits.

Tags created through the GitHub interface (or pushed from a different location) will be available
locally the next time you `git fetch`.

Keep in mind that the GitHub repository page is one of the main entry points for new users of your package.
Your project looks healthier and better maintained with a nice, well documented history of
[releases](https://docs.github.com/en/repositories/releasing-projects-on-github/viewing-your-repositorys-releases-and-tags).
~~~
{: .callout}
```

Tags and GitHub releases do not automatically make your package easier to find or install,
but they can be helpful in preparing the source of version information or the source archive
as you prepare the *distribution* of your package.

**Comment by eirrgang** [comment]
> I'd like to go through the distribution lesson a bit more. It's not clear to me right now which points are informational and which are tutorial (what steps should be executed when). Would like to keep hands on (test-pypi) upfront, then we can put information about conda at the end.

Are there any changes you want to request or suggest before merging to the staging branch? Or should we merge to `2022-review` and then propose changes there?

**Pull request merged**
